app/src/main/java/org/onionshare/android/ui/MainViewModel.kt

```diff
@@ -22,41 +27,64 @@ private val LOG = LoggerFactory.getLogger(MainViewModel::class.java)
     private val fileManager: FileManager,
 ) : AndroidViewModel(app) {
 
-    private val _fileManagerState = MutableStateFlow<FileManager.State>(FileManager.State.NoFiles)
-    val fileManagerState: StateFlow<FileManager.State> = _fileManagerState
-    val webserverState = webserverManager.state
+    private val _shareState = MutableStateFlow<ShareUiState>(ShareUiState.NoFiles)
+    val shareState: StateFlow<ShareUiState> = _shareState
+
+    @Volatile
+    var startSharingJob: Job? = null
 
     override fun onCleared() {
         super.onCleared()
         // TODO later, we shouldn't stop the server when closing the activity
-        stopServer()
+        stopSharing()
     }
 
     fun onUrisReceived(uris: List<Uri>) {
         if (uris.isEmpty()) return // user backed out of select activity
 
         // not supporting selecting entire folders with sub-folders
         viewModelScope.launch(Dispatchers.IO) {
-            val filesAdded = fileManager.addFiles(uris, fileManagerState.value)
-            _fileManagerState.value = filesAdded
+            val filesAdded = fileManager.addFiles(uris, shareState.value.files)
+            val totalSize = filesAdded.files.sumOf { it.size }
+            _shareState.value = ShareUiState.FilesAdded(filesAdded.files, totalSize)
         }
     }
 
-    fun startServer() {
-        // FIXME this is a mixing of concerns to get the right state in the UI
-        webserverManager.onFilesBeingZipped()
+    fun removeFile(file: SendFile) {
+        val newList = shareState.value.files.filterNot { it == file }
+        if (newList.isEmpty()) {
+            _shareState.value = ShareUiState.NoFiles
+        } else {
+            val totalSize = newList.sumOf { it.size }
+            _shareState.value = ShareUiState.FilesAdded(newList, totalSize)
+        }
+    }
 
-        viewModelScope.launch(Dispatchers.IO) {
-            val filesAdded = fileManagerState.value as FileManager.State.FilesAdded
-            val filesReady = fileManager.zipFiles(filesAdded)
-            _fileManagerState.value = filesReady
+    fun removeAll() {
+        _shareState.value = ShareUiState.NoFiles
+    }
+
+    fun onSheetButtonClicked() {
+        when (shareState.value) {
+            is ShareUiState.FilesAdded -> startSharing()
+            is ShareUiState.Starting -> stopSharing()
+            is ShareUiState.Sharing -> stopSharing()
+            is ShareUiState.Complete -> startSharing()
+            ShareUiState.NoFiles -> error("Sheet button should not be visible with no files")
+        }
+    }
 
-            LOG.error("$filesReady")
+    private fun startSharing() {
+        startSharingJob = viewModelScope.launch(Dispatchers.IO) {
+            val files = shareState.value.files
+            _shareState.value = ShareUiState.Starting(files, shareState.value.totalSize)
+            // call ensureActive() before any heavy work to ensure we don't continue when cancelled
+            ensureActive()
```

Comment:
If we've already set _shareState.value before checking this, is it possible that we've overwritten the new cancelled state with the starting state?

Reply:
You mean that after starting to share, the user presses so quickly again that we get to the end of `stopSharing()` where  new state gets set before we set the starting state here? I'd say unlikely, but maybe not impossible. Would it help to call ensureActive() as well before setting the state?

Btw. I consider all this state handling here prototypy and just did it also with cancellation to show how it *could* be done with coroutines in Kotlin, but maybe we need to overhaul it completely before release.